Makefile

```diff
@@ -820,19 +820,24 @@ enter:
 grpc:
 	$(MAKE) -C build.assets grpc
 
-# proto file dependencies within the api module must be passed with the 'M' flag. This
-# way protoc generated files will use the correct api module import path in the case where
-# the import path has a version suffix, e.g. github.com/gravitational/teleport/api/v8
-GOGOPROTO_IMPORTMAP ?= $\
-	Mgithub.com/gravitational/teleport/api/types/types.proto=$(API_IMPORT_PATH)/types,$\
-	Mgithub.com/gravitational/teleport/api/types/events/events.proto=$(API_IMPORT_PATH)/types/events,$\
-	Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=$(API_IMPORT_PATH)/types/wrappers,$\
-	Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=$(API_IMPORT_PATH)/types/webauthn
+print/env:
+	env
 
 # buildbox-grpc generates GRPC stubs
 .PHONY: buildbox-grpc
+buildbox-grpc: API_IMPORT_PATH := $(shell head -1 api/go.mod | awk '{print $$2}')
+# Proto file dependencies within the api module must be passed with the 'M'
+# flag. This way protoc generated files will use the correct api module import
+# path in the case where the import path has a version suffix, e.g.
+# "github.com/gravitational/teleport/api/v8".
+buildbox-grpc: GOGOPROTO_IMPORTMAP := $\
+	Mgithub.com/gravitational/teleport/api/types/events/events.proto=$(API_IMPORT_PATH)/types/events,$\
+	Mgithub.com/gravitational/teleport/api/types/types.proto=$(API_IMPORT_PATH)/types,$\
+	Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=$(API_IMPORT_PATH)/types/webauthn,$\
+	Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=$(API_IMPORT_PATH)/types/wrappers,$\
+	Mignoreme=ignoreme
 buildbox-grpc:
```

Comment:
What are the semantics/purpose of having multiple `.PHONY` targets of the same name? I haven't been able to find good information on that online.

Might it work to just make this
```Makefile
# buildbox-grpc generates GRPC stubs
.PHONY: buildbox-grpc
buildbox-grpc:
	API_IMPORT_PATH := $(shell head -1 api/go.mod | awk '{print $$2}')

# Proto file dependencies within the api module must be passed with the 'M'
# flag. This way protoc generated files will use the correct api module import
# path in the case where the import path has a version suffix, e.g.
# "github.com/gravitational/teleport/api/v8".
	GOGOPROTO_IMPORTMAP := $\
	Mgithub.com/gravitational/teleport/api/types/events/events.proto=$(API_IMPORT_PATH)/types/events,$\
	Mgithub.com/gravitational/teleport/api/types/types.proto=$(API_IMPORT_PATH)/types,$\
	Mgithub.com/gravitational/teleport/api/types/webauthn/webauthn.proto=$(API_IMPORT_PATH)/types/webauthn,$\
	Mgithub.com/gravitational/teleport/api/types/wrappers/wrappers.proto=$(API_IMPORT_PATH)/types/wrappers,$\
	Mignoreme=ignoreme

	@echo "PROTO_INCLUDE = $$PROTO_INCLUDE"
	$(CLANG_FORMAT) -i -style='{ColumnLimit: 100, IndentWidth: 4, Language: Proto}' \
		api/client/proto/authservice.proto \
		api/client/proto/joinservice.proto \
		api/types/events/events.proto \
		api/types/types.proto \
		api/types/webauthn/webauthn.proto \
		api/types/wrappers/wrappers.proto \
		lib/datalog/types.proto \
		lib/events/slice.proto \
		lib/multiplexer/test/ping.proto \
		lib/web/envelope.proto

	cd api/client/proto && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		certs.proto authservice.proto joinservice.proto

	cd api/types/events && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		events.proto

	cd api/types && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		types.proto

	cd api/types/webauthn && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		webauthn.proto

	cd api/types/wrappers && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		wrappers.proto

	cd lib/datalog && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		types.proto

	cd lib/events && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		slice.proto

	cd lib/multiplexer/test && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		ping.proto

	cd lib/web && protoc -I=.:$$PROTO_INCLUDE \
		--gogofast_out=plugins=grpc,$(GOGOPROTO_IMPORTMAP):. \
		envelope.proto

```

Reply:
These are just per-target variable overrides. It does what your snippet is trying to do.

I don't think your snippet works, though. We can't define make-like variable in the target's body.

Reply:
It's worth noting that target-specific variables like this one are initialized on every call to `make`, whereas setting the variable within the target like Isaiah is meaning to ( which can be done with $(eval API_IMPORT_PATH := $(shell head -1 api/go.mod | awk '{print $$2}')) ) would only be initialized in the specified target. You can see this is true if you add `shell && sleep 5 ...`. I'm not sure which approach is actually best practice for that reason, though here the performance impact is negligible.

Reply:
Yep. It should be negligible in this case, so I'd rather take the cleaner syntax to the evals (which I think are less usual).